src/psyclone/psyGen.py

```diff
@@ -2659,7 +2669,28 @@ def _get_private_list(self):
             # We have at least two accesses. If the first one is a write,
             # assume the variable should be private:
             if accesses[0].access_type == AccessType.WRITE:
-                result.add(var_name.lower())
+                # Check if the write access is inside the parallel loop. If
+                # the write is outside of a loop, it is an assignment to
+                # a shared variable. Example where jpk is likely used
+                # outside of the parallel section later, so it must be
+                # declared as shared in order to have its value in other loops:
+                # !$omp parallel
+                # jpk = 100
```

Comment:
Wouldn't this assignment cause us problems anyway as one thread could attempt to read `jpk` while another writes to it? We should really protect it with an `atomic` or some such?

Reply:
Or suggest to the user that they move the assignment?

Reply:
Yes, the handling of scalars is a HUGE problem, e.g.:
```
!$omp parallel
   a=100
   b=100
   do i=1, b
     x(i) = a*i
   end do
!$omp end parallel
...
    do i=1, b
      a = x(i)*x(i)
      y(i) = sin(a)
```
In this example, b must be shared, but you can only know that when you see the read access in the 2nd loop. a could be private or shared, but if the second loop is included in the same parallel do directive, it must be private.

Since an assignment is only accepted if explicitly disabling the node-type check (and since this issue is not really related to the InlineKernel (except that the previous shared declaration of b as shared stopped working), I suggest to leave this for now as an issue the user has to handle (by explicitly disabling the node test). I am happy to open a separate issue.

Yes, technically I as a user would have to add an atomic, but since all threads store the same value (and each thread only reads after it has stored this value), the variable should at all times have the correct value (and yes, I have been bitten by a similar assignment, that the compiler decided to implement as 2 4-byte stores once, but in that case the values where not the same between all threads).

Reply:
As discussed, please could you open an Issue for this (or check that we don't already have one).

Reply:
Done: #598.